File: internal/csvons/fieldexpr_test.go
package csvons

import (
	"testing"
)

// TestPlainField_FieldValue verifies that PlainField correctly extracts
// values from a single column across all data rows.
func TestPlainField_FieldValue(t *testing.T) {
	metadata := &Metadata{
		DataIndex: 1,
	}

	field := &PlainField{
		metadata:  metadata,
		fieldName: "field1",
	}

	fields := []string{"field1", "field2"}
	records := [][]string{
		{"header1", "header2"},
		{"value1", "value2"},
		{"value3", "value4"},
	}

	ch := field.FieldValue(fields, records)
	var results []string
	for val := range ch {
		results = append(results, val)
	}

	expected := []string{"value1", "value3"}
	if len(results) != len(expected) {
		t.Errorf("PlainField.FieldValue() returned %d values, expected %d", len(results), len(expected))
	}
	for i, val := range results {
		if val != expected[i] {
			t.Errorf("PlainField.FieldValue() result[%d] = %v, expected %v", i, val, expected[i])
		}
	}
}

// TestPlainField_typeString verifies the type identifier string for PlainField.
func TestPlainField_typeString(t *testing.T) {
	field := &PlainField{}
	result := field.typeString()
	expected := "plain"
	if result != expected {
		t.Errorf("PlainField.typeString() = %v, expected %v", result, expected)
	}
}

// TestPlainField_Init verifies that Init correctly sets metadata and field name.
func TestPlainField_Init(t *testing.T) {
	metadata := &Metadata{DataIndex: 1}
	field := &PlainField{}
	field.Init(metadata, "testField")

	if field.metadata != metadata {
		t.Errorf("PlainField.Init() metadata not set correctly")
	}
	if field.fieldName != "testField" {
		t.Errorf("PlainField.Init() fieldName = %v, expected testField", field.fieldName)
	}
}

// TestRepeatField_FieldValue verifies that RepeatField correctly splits
// cell values by the level-1 separator and yields each element.
func TestRepeatField_FieldValue(t *testing.T) {
	metadata := &Metadata{
		DataIndex:     1,
		Lev1Separator: ",",
	}

	field := &RepeatField{
		metadata:  metadata,
		fieldName: "field1",
	}

	fields := []string{"field1", "field2"}
	records := [][]string{
		{"header1", "header2"},
		{"a,b,c", "value2"},
		{"x,y", "value4"},
	}

	ch := field.FieldValue(fields, records)
	var results []string
	for val := range ch {
		results = append(results, val)
	}

	expected := []string{"a", "b", "c", "x", "y"}
	if len(results) != len(expected) {
		t.Errorf("RepeatField.FieldValue() returned %d values, expected %d", len(results), len(expected))
	}
	for i, val := range results {
		if val != expected[i] {
			t.Errorf("RepeatField.FieldValue() result[%d] = %v, expected %v", i, val, expected[i])
		}
	}
}

// TestRepeatField_typeString verifies the type identifier string for RepeatField.
func TestRepeatField_typeString(t *testing.T) {
	field := &RepeatField{}
	result := field.typeString()
	expected := "repeat"
	if result != expected {
		t.Errorf("RepeatField.typeString() = %v, expected %v", result, expected)
	}
}

// TestRepeatField_Init verifies that Init correctly strips the "[]" suffix.
func TestRepeatField_Init(t *testing.T) {
	metadata := &Metadata{DataIndex: 1}
	field := &RepeatField{}
	field.Init(metadata, "testField[]")

	if field.metadata != metadata {
		t.Errorf("RepeatField.Init() metadata not set correctly")
	}
	if field.fieldName != "testField" {
		t.Errorf("RepeatField.Init() fieldName = %v, expected testField", field.fieldName)
	}
}

// TestNestedField_FieldValue verifies that NestedField correctly splits
// by both separators and extracts the value at the specified index.
func TestNestedField_FieldValue(t *testing.T) {
	metadata := &Metadata{
		DataIndex:     1,
		Lev1Separator: ",",
		Lev2Separator: ":",
	}

	field := &NestedField{
		metadata:  metadata,
		fieldName: "field1",
		index:     1,
	}

	fields := []string{"field1", "field2"}
	records := [][]string{
		{"header1", "header2"},
		{"a:b:c,d:e:f", "value2"},
		{"x:y:z", "value4"},
	}

	ch := field.FieldValue(fields, records)
	var results []string
	for val := range ch {
		results = append(results, val)
	}

	expected := []string{"b", "e", "y"}
	if len(results) != len(expected) {
		t.Errorf("NestedField.FieldValue() returned %d values, expected %d", len(results), len(expected))
	}
	for i, val := range results {
		if val != expected[i] {
			t.Errorf("NestedField.FieldValue() result[%d] = %v, expected %v", i, val, expected[i])
		}
	}
}

// TestNestedField_typeString verifies the type identifier string for NestedField.
func TestNestedField_typeString(t *testing.T) {
	field := &NestedField{}
	result := field.typeString()
	expected := "nested"
	if result != expected {
		t.Errorf("NestedField.typeString() = %v, expected %v", result, expected)
	}
}

// TestNestedField_Init verifies that Init correctly parses the field name and index.
func TestNestedField_Init(t *testing.T) {
	metadata := &Metadata{DataIndex: 1}
	field := &NestedField{}
	field.Init(metadata, "testField{2}")

	if field.metadata != metadata {
		t.Errorf("NestedField.Init() metadata not set correctly")
	}
	if field.fieldName != "testField" {
		t.Errorf("NestedField.Init() fieldName = %v, expected testField", field.fieldName)
	}
	if field.index != 2 {
		t.Errorf("NestedField.Init() index = %v, expected 2", field.index)
	}
}

// TestComplexField_FieldValue verifies that ComplexField correctly concatenates
// values from multiple columns using the field connector.
func TestComplexField_FieldValue(t *testing.T) {
	metadata := &Metadata{
		DataIndex:      1,
		FieldConnector: "-",
	}

	field := &ComplexField{
		metadata:   metadata,
		fieldNames: []string{"field1", "field2"},
	}

	fields := []string{"field1", "field2", "field3"}
	records := [][]string{
		{"header1", "header2", "header3"},
		{"value1", "value2", "value3"},
		{"value4", "value5", "value6"},
	}

	ch := field.FieldValue(fields, records)
	var results []string
	for val := range ch {
		results = append(results, val)
	}

	expected := []string{"value1-value2-", "value4-value5-"}
	if len(results) != len(expected) {
		t.Errorf("ComplexField.FieldValue() returned %d values, expected %d", len(results), len(expected))
	}
	for i, val := range results {
		if val != expected[i] {
			t.Errorf("ComplexField.FieldValue() result[%d] = %v, expected %v", i, val, expected[i])
		}
	}
}

// TestComplexField_typeString verifies the type identifier string for ComplexField.
func TestComplexField_typeString(t *testing.T) {
	field := &ComplexField{}
	result := field.typeString()
	expected := "complex"
	if result != expected {
		t.Errorf("ComplexField.typeString() = %v, expected %v", result, expected)
	}
}

// TestComplexField_Init verifies that Init correctly parses field names from curly braces.
func TestComplexField_Init(t *testing.T) {
	metadata := &Metadata{DataIndex: 1}
	field := &ComplexField{}
	field.Init(metadata, "{field1}{field2}")

	if field.metadata != metadata {
		t.Errorf("ComplexField.Init() metadata not set correctly")
	}
	expected := []string{"field1", "field2"}
	if len(field.fieldNames) != len(expected) {
		t.Errorf("ComplexField.Init() fieldNames length = %d, expected %d", len(field.fieldNames), len(expected))
	}
	for i, name := range field.fieldNames {
		if name != expected[i] {
			t.Errorf("ComplexField.Init() fieldNames[%d] = %v, expected %v", i, name, expected[i])
		}
	}
}

// TestGenerateFieldExpr uses table-driven tests to verify that GenerateFieldExpr
// correctly identifies and creates the appropriate FieldExpr type for each expression pattern.
func TestGenerateFieldExpr(t *testing.T) {
	metadata := &Metadata{
		DataIndex:      1,
		Lev1Separator:  ",",
		Lev2Separator:  ":",
		FieldConnector: "-",
	}

	tests := []struct {
		name         string
		fieldExpr    string
		expectedType string
		shouldBeNil  bool
	}{
		{
			name:         "Plain field expression",
			fieldExpr:    "field1",
			expectedType: "plain",
			shouldBeNil:  false,
		},
		{
			name:         "Plain field with numbers",
			fieldExpr:    "field123",
			expectedType: "plain",
			shouldBeNil:  false,
		},
		{
			name:         "Repeat field expression",
			fieldExpr:    "field1[]",
			expectedType: "repeat",
			shouldBeNil:  false,
		},
		{
			name:         "Repeat field with numbers",
			fieldExpr:    "field123[]",
			expectedType: "repeat",
			shouldBeNil:  false,
		},
		{
			name:         "Nested field expression",
			fieldExpr:    "field1{0}",
			expectedType: "nested",
			shouldBeNil:  false,
		},
		{
			name:         "Nested field with numbers",
			fieldExpr:    "field123{2}",
			expectedType: "nested",
			shouldBeNil:  false,
		},
		{
			name:         "Complex field expression single",
			fieldExpr:    "{field1}",
			expectedType: "complex",
			shouldBeNil:  false,
		},
		{
			name:         "Complex field expression multiple",
			fieldExpr:    "{field1}{field2}",
			expectedType: "",
			shouldBeNil:  true,
		},
		{
			name:         "Invalid field expression",
			fieldExpr:    "field-1",
			expectedType: "",
			shouldBeNil:  true,
		},
		{
			name:         "Empty field expression",
			fieldExpr:    "",
			expectedType: "",
			shouldBeNil:  true,
		},
		{
			name:         "Field with special characters",
			fieldExpr:    "field@1",
			expectedType: "",
			shouldBeNil:  true,
		},
		{
			name:         "Mixed case field",
			fieldExpr:    "Field1",
			expectedType: "plain",
			shouldBeNil:  false,
		},
		{
			name:         "Mixed case repeat field",
			fieldExpr:    "Field1[]",
			expectedType: "repeat",
			shouldBeNil:  false,
		},
		{
			name:         "Mixed case nested field",
			fieldExpr:    "Field1{1}",
			expectedType: "nested",
			shouldBeNil:  false,
		},
		{
			name:         "Mixed case complex field",
			fieldExpr:    "{Field1}",
			expectedType: "complex",
			shouldBeNil:  false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := GenerateFieldExpr(metadata, tt.fieldExpr)

			if tt.shouldBeNil {
				if result != nil {
					t.Errorf("GenerateFieldExpr(%v) = %v, expected nil", tt.fieldExpr, result)
				}
				return
			}

			if result == nil {
				t.Errorf("GenerateFieldExpr(%v) = nil, expected non-nil", tt.fieldExpr)
				return
			}

			actualType := result.typeString()
			if actualType != tt.expectedType {
				t.Errorf("GenerateFieldExpr(%v).typeString() = %v, expected %v", tt.fieldExpr, actualType, tt.expectedType)
			}
		})
	}
}

// TestGenerateFieldExpr_NilMetadata tests the nil metadata case.
// Skipped because log.Fatal will exit the program in a unit test environment.
func TestGenerateFieldExpr_NilMetadata(t *testing.T) {
	t.Skip("Skipping nil metadata test - log.Fatal exits program")
}

// TestFieldExprInterface verifies that all field types correctly implement
// the FieldExpr interface at compile time.
func TestFieldExprInterface(t *testing.T) {
	var _ FieldExpr = (*PlainField)(nil)
	var _ FieldExpr = (*RepeatField)(nil)
	var _ FieldExpr = (*NestedField)(nil)
	var _ FieldExpr = (*ComplexField)(nil)
}

// TestFieldExprMap verifies that fieldExprMap contains the expected regex patterns
// and that each constructor creates an instance of the correct type.
func TestFieldExprMap(t *testing.T) {
	expectedPatterns := []string{
		`^[a-zA-Z0-9]+$`,
		`^[a-zA-Z0-9]+\[\]$`,
		`^[a-zA-Z0-9]+\{\d+\}$`,
		`^\{[a-zA-Z0-9]+\}+$`,
	}

	expectedTypes := []string{
		"plain",
		"repeat",
		"nested",
		"complex",
	}

	for i, pattern := range expectedPatterns {
		constructor, exists := fieldExprMap[pattern]
		if !exists {
			t.Errorf("fieldExprMap missing pattern: %v", pattern)
			continue
		}

		// Verify the constructor creates the right type.
		instance := constructor("test")
		actualType := instance.typeString()
		if actualType != expectedTypes[i] {
			t.Errorf("fieldExprMap[%v] creates type %v, expected %v", pattern, actualType, expectedTypes[i])
		}
	}
}

// TestFieldTypesWithEmptyRecords verifies that all field types handle
// empty records gracefully by returning zero values through the channel.
func TestFieldTypesWithEmptyRecords(t *testing.T) {
	metadata := &Metadata{DataIndex: 1}
	emptyRecords := [][]string{}

	plainField := &PlainField{metadata: metadata, fieldName: "field1"}
	repeatField := &RepeatField{metadata: metadata, fieldName: "field1"}
	nestedField := &NestedField{metadata: metadata, fieldName: "field1", index: 0}
	complexField := &ComplexField{metadata: metadata, fieldNames: []string{"field1"}}

	fields := []string{"field1"}

	tests := []struct {
		name   string
		field  FieldExpr
		expect int // Expected number of output values.
	}{
		{"PlainField with empty records", plainField, 0},
		{"RepeatField with empty records", repeatField, 0},
		{"NestedField with empty records", nestedField, 0},
		{"ComplexField with empty records", complexField, 0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			ch := tt.field.FieldValue(fields, emptyRecords)
			count := 0
			for range ch {
				count++
			}
			if count != tt.expect {
				t.Errorf("%s.FieldValue() returned %d values, expected %d", tt.name, count, tt.expect)
			}
		})
	}
}

// TestFieldTypesWithNilRecords verifies that all field types handle
// nil records gracefully by returning zero values through the channel.
func TestFieldTypesWithNilRecords(t *testing.T) {
	metadata := &Metadata{DataIndex: 1}
	fields := []string{"field1"}

	plainField := &PlainField{metadata: metadata, fieldName: "field1"}
	repeatField := &RepeatField{metadata: metadata, fieldName: "field1"}
	nestedField := &NestedField{metadata: metadata, fieldName: "field1", index: 0}
	complexField := &ComplexField{metadata: metadata, fieldNames: []string{"field1"}}

	tests := []struct {
		name   string
		field  FieldExpr
		expect int // Expected number of output values.
	}{
		{"PlainField with nil records", plainField, 0},
		{"RepeatField with nil records", repeatField, 0},
		{"NestedField with nil records", nestedField, 0},
		{"ComplexField with nil records", complexField, 0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			ch := tt.field.FieldValue(fields, nil)
			count := 0
			for range ch {
				count++
			}
			if count != tt.expect {
				t.Errorf("%s.FieldValue() returned %d values, expected %d", tt.name, count, tt.expect)
			}
		})
	}
}

// TestFieldNotFound verifies that all field types return nil when
// the specified field is not present in the column headers.
func TestFieldNotFound(t *testing.T) {
	metadata := &Metadata{DataIndex: 1}
	fields := []string{"otherField"}
	records := [][]string{
		{"header1"},
		{"value1"},
	}

	plainField := &PlainField{metadata: metadata, fieldName: "field1"}
	repeatField := &RepeatField{metadata: metadata, fieldName: "field1"}
	nestedField := &NestedField{metadata: metadata, fieldName: "field1", index: 0}
	complexField := &ComplexField{metadata: metadata, fieldNames: []string{"field1"}}

	tests := []struct {
		name  string
		field FieldExpr
	}{
		{"PlainField field not found", plainField},
		{"RepeatField field not found", repeatField},
		{"NestedField field not found", nestedField},
		{"ComplexField field not found", complexField},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			ch := tt.field.FieldValue(fields, records)
			if ch != nil {
				t.Errorf("%s.FieldValue() should return nil when field not found", tt.name)
			}
		})
	}
}

File: gui/electron/test/core/field-expr.test.ts
import { describe, it, expect } from 'vitest';
import {
  PlainField,
  RepeatField,
  NestedField,
  ComplexField,
  generateFieldExpr,
} from '../../src/core/field-expr.ts';
import type { Metadata } from '../../src/core/types.ts';

const baseMetadata: Metadata = {
  csv_file_folder: '',
  name_index: 0,
  data_index: 1,
  extension: '.csv',
  lev1_separator: ',',
  lev2_separator: ':',
  field_connector: '-',
};

describe('PlainField', () => {
  it('extracts values from a single column', () => {
    const field = new PlainField();
    field.init(baseMetadata, 'field1');

    const fields = ['field1', 'field2'];
    const records = [
      ['header1', 'header2'],
      ['value1', 'value2'],
      ['value3', 'value4'],
    ];

    const results = [...field.fieldValues(fields, records)!];
    expect(results).toEqual(['value1', 'value3']);
  });

  it('returns "plain" for typeString', () => {
    expect(new PlainField().typeString()).toBe('plain');
  });

  it('init sets metadata and fieldName', () => {
    const field = new PlainField();
    field.init(baseMetadata, 'testField');
    expect(field.metadata).toBe(baseMetadata);
    expect(field.fieldName).toBe('testField');
  });

  it('returns empty for field not found', () => {
    const field = new PlainField();
    field.init(baseMetadata, 'field1');
    const results = field.fieldValues(['otherField'], [['h'], ['v']]);
    expect([...results!]).toEqual([]);
  });

  it('returns empty for empty records', () => {
    const field = new PlainField();
    field.init(baseMetadata, 'field1');
    const results = [...field.fieldValues(['field1'], [])!];
    expect(results).toEqual([]);
  });
});

describe('RepeatField', () => {
  it('splits cell values by lev1_separator', () => {
    const field = new RepeatField();
    field.init(baseMetadata, 'field1[]');

    const fields = ['field1', 'field2'];
    const records = [
      ['header1', 'header2'],
      ['a,b,c', 'value2'],
      ['x,y', 'value4'],
    ];

    const results = [...field.fieldValues(fields, records)!];
    expect(results).toEqual(['a', 'b', 'c', 'x', 'y']);
  });

  it('returns "repeat" for typeString', () => {
    expect(new RepeatField().typeString()).toBe('repeat');
  });

  it('init strips [] suffix', () => {
    const field = new RepeatField();
    field.init(baseMetadata, 'testField[]');
    expect(field.fieldName).toBe('testField');
  });
});

describe('NestedField', () => {
  it('extracts values at nested index', () => {
    const field = new NestedField();
    field.init(baseMetadata, 'field1{1}');

    const fields = ['field1', 'field2'];
    const records = [
      ['header1', 'header2'],
      ['a:b:c,d:e:f', 'value2'],
      ['x:y:z', 'value4'],
    ];

    const results = [...field.fieldValues(fields, records)!];
    expect(results).toEqual(['b', 'e', 'y']);
  });

  it('returns "nested" for typeString', () => {
    expect(new NestedField().typeString()).toBe('nested');
  });

  it('init parses fieldName and index', () => {
    const field = new NestedField();
    field.init(baseMetadata, 'testField{2}');
    expect(field.fieldName).toBe('testField');
    expect(field.index).toBe(2);
  });
});

describe('ComplexField', () => {
  it('concatenates values from multiple columns', () => {
    const field = new ComplexField();
    field.init(baseMetadata, '{field1}{field2}');

    const fields = ['field1', 'field2', 'field3'];
    const records = [
      ['header1', 'header2', 'header3'],
      ['value1', 'value2', 'value3'],
      ['value4', 'value5', 'value6'],
    ];

    const results = [...field.fieldValues(fields, records)!];
    expect(results).toEqual(['value1-value2-', 'value4-value5-']);
  });

  it('returns "complex" for typeString', () => {
    expect(new ComplexField().typeString()).toBe('complex');
  });

  it('init parses field names from braces', () => {
    const field = new ComplexField();
    field.init(baseMetadata, '{field1}{field2}');
    expect(field.fieldNames).toEqual(['field1', 'field2']);
  });

  it('returns empty when any field index is out of bounds for a record', () => {
    const field = new ComplexField();
    field.init(baseMetadata, '{field1}{field2}');

    // field1 is at index 0, field2 is at index 1
    // row with only 1 element: fieldIndex=1 >= record.length=1 → valid=false
    const fields = ['field1', 'field2'];
    const records = [
      ['header1', 'header2'],
      ['onlyone'],           // record.length=1, fieldIndex for field2=1 → out of bounds
      ['val1', 'val2'],     // valid row
    ];

    const results = [...field.fieldValues(fields, records)!];
    // only the valid row should be yielded
    expect(results).toEqual(['val1-val2-']);
  });

  it('returns empty when a required field name is not in fields list', () => {
    const field = new ComplexField();
    field.init(baseMetadata, '{field1}{missing}');

    const fields = ['field1', 'field2'];
    const records = [
      ['header1', 'header2'],
      ['value1', 'value2'],
    ];

    // missing field causes early return (fieldIndexes.some idx===-1)
    const results = [...field.fieldValues(fields, records)!];
    expect(results).toEqual([]);
  });
});

describe('generateFieldExpr', () => {
  const tests = [
    { name: 'Plain field', expr: 'field1', type: 'plain', nil: false },
    { name: 'Plain with numbers', expr: 'field123', type: 'plain', nil: false },
    { name: 'Repeat field', expr: 'field1[]', type: 'repeat', nil: false },
    { name: 'Repeat with numbers', expr: 'field123[]', type: 'repeat', nil: false },
    { name: 'Nested field', expr: 'field1{0}', type: 'nested', nil: false },
    { name: 'Nested with numbers', expr: 'field123{2}', type: 'nested', nil: false },
    { name: 'Complex single', expr: '{field1}', type: 'complex', nil: false },
    { name: 'Complex multiple', expr: '{field1}{field2}', type: '', nil: true },
    { name: 'Invalid with dash', expr: 'field-1', type: '', nil: true },
    { name: 'Empty string', expr: '', type: '', nil: true },
    { name: 'Special chars', expr: 'field@1', type: '', nil: true },
    { name: 'Mixed case', expr: 'Field1', type: 'plain', nil: false },
    { name: 'Mixed case repeat', expr: 'Field1[]', type: 'repeat', nil: false },
    { name: 'Mixed case nested', expr: 'Field1{1}', type: 'nested', nil: false },
    { name: 'Mixed case complex', expr: '{Field1}', type: 'complex', nil: false },
  ];

  for (const tt of tests) {
    it(tt.name, () => {
      const result = generateFieldExpr(baseMetadata, tt.expr);
      if (tt.nil) {
        expect(result).toBeNull();
      } else {
        expect(result).not.toBeNull();
        expect(result!.typeString()).toBe(tt.type);
      }
    });
  }

  it('throws on null metadata', () => {
    expect(() => generateFieldExpr(null as any, 'field1')).toThrow(
      'metadata is nil',
    );
  });
});
